src/ejson2/EjsonDecoder.ts

```diff
@@ -0,0 +1,377 @@
+import {
+  BsonBinary,
+  BsonDbPointer,
+  BsonDecimal128,
+  BsonFloat,
+  BsonInt32,
+  BsonInt64,
+  BsonJavascriptCode,
+  BsonJavascriptCodeWithScope,
+  BsonMaxKey,
+  BsonMinKey,
+  BsonObjectId,
+  BsonSymbol,
+  BsonTimestamp,
+} from '../bson/values';
+
+export interface EjsonDecoderOptions {
+  /** Whether to parse legacy Extended JSON formats */
+  legacy?: boolean;
+}
+
+export class EjsonDecoder {
+  constructor(private options: EjsonDecoderOptions = {}) {}
+
+  public decode(json: string): unknown {
+    const parsed = JSON.parse(json);
```

Comment:
Do not use `JSON.parse()` read from the binary "reader" directly. Follow the pattern used in `JsonDecoder`.